File: python_api/SparkVersion/Operations/CreateModelOperation.py
import time

from pyspark.ml.feature import VectorAssembler, StringIndexer, StandardScaler, MinMaxScaler
from pyspark.sql.types import StringType

import Context
from SparkVersion.Operations.Helpers.ReadWriteFiler import readData
from SparkVersion.ml.ClassificatorsAndRegressors import ClassificatorsAndRegressors


def createModel(params):
    data = readData()
    start = time.time()

    preparedData = extractData(data, params['label'], params['features'], params['options']['standardization'])
    (train, test) = preparedData.randomSplit([params['options']['trainSize'], 1-params['options']['trainSize']])
    prepareDataTime = time.time() - start

    mlProblem = ClassificatorsAndRegressors(preparedData, train, test)\
        .chooseTypeOfProblem(params['options']['typeOfProblem'])

    mlModel = mlProblem.chooseMethod(params['options']['method'], params['hyperparameters'])

    prediction = mlProblem.fitAndTransform(mlModel)

    trainingModelTime = time.time() - start - prepareDataTime
    result = mlProblem.calculateMetrics(prediction)
    calculateMetricsTime = time.time() - start - trainingModelTime - prepareDataTime

    result.setPrepareDataTime(time=prepareDataTime)
    result.setTrainingModelTime(time=trainingModelTime)
    result.setCalculateMetricsTime(time=calculateMetricsTime)

    return result


def extractData(data, label, features, standardization):
    def setIndexesName(name):
        return name + "_index"

    allTextFutures = [field.name for field in data.schema.fields if (isinstance(field.dataType, StringType))]
    textFeatures = [name for name in features if name in allTextFutures]
    newNameOfTextFeatures = [setIndexesName(name) for name in textFeatures]
    numberFeature = [feature for feature in features if (feature not in textFeatures)]
    newFeatures = numberFeature + newNameOfTextFeatures

    textFeaturesIndexer = StringIndexer(
        inputCols=textFeatures,
        outputCols=newNameOfTextFeatures
    ).fit(data)
    if Context.saveModels:
        textFeaturesIndexer.write().overwrite().save(Context.getTextFeatureIndexerFilePath())
    newData = textFeaturesIndexer.transform(data)
    vector = VectorAssembler(
        inputCols=newFeatures,
        outputCol="nonScaledFeatures"
    ).transform(newData).withColumnRenamed(label, "label").select("nonScaledFeatures", "label")
    if standardization == "StandardScaler":
        sScaler = StandardScaler(withMean=True, withStd=True, inputCol="nonScaledFeatures", outputCol="features")
        return sScaler.fit(vector).transform(vector).select("features", "label")
    elif standardization == "MinMaxScaler":
        mmScaler = MinMaxScaler(inputCol="nonScaledFeatures", outputCol="features")
        return mmScaler.fit(vector).transform(vector).select("features", "label")
    else:
        return vector.withColumnRenamed("nonScaledFeatures", "features")

File: scala_api/src/main/scala/application/dataAnalysis/operations/CreateModelOperation.scala
package application.dataAnalysis.operations

import application.dataAnalysis.Context
import application.dataAnalysis.Context.getTextFeatureIndexerFilePath
import application.dataAnalysis.ml.ClassificatorsAndRegressors
import application.dataAnalysis.operations.Helpers.ReadWriteFiler.readData
import application.dataAnalysis.operations.Helpers.StatisticsCreator
import application.models.MachineLearningModelParams
import application.models.SummaryOfMerics.SummaryResult
import org.apache.spark.ml.feature._
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.types.{StringType, StructField}

object CreateModelOperation extends StatisticsCreator {

  def apply(params: MachineLearningModelParams): SummaryResult = {
    val data = readData()
    val start = System.currentTimeMillis()

    val preparedData = extractData(data, params.label, params.features, params.options.standardization.get)
    val Array(train, test) = preparedData.randomSplit(params.options.division)
    val prepareDataTime = System.currentTimeMillis() - start

    val mlModel= new ClassificatorsAndRegressors(preparedData, train, test)
      .chooseTypeOfProblem(params.options.typeOfProblem)

    val prediction = mlModel
      .predicate(params.options.method, params.hyperparameters)
    val trainingModelTime = System.currentTimeMillis() - start - prepareDataTime

    val result = mlModel.calculateMetrics(prediction)
    val calculateMetricsTime = System.currentTimeMillis() - start - trainingModelTime - prepareDataTime

    result.copy(
      prepareDataTime = prepareDataTime.toFloat / 1000,
      trainingModelTime = trainingModelTime.toFloat / 1000,
      calculateMetricsTime= calculateMetricsTime.toFloat / 1000
    )
  }

  private def extractData(data: DataFrame, label: String, features: Array[String], standardization: String): DataFrame = {
    val setIndexesName: String => String = name => s"${name}_index"

    val textFeatures: Array[String] = features.filter(parameter => {
        data.schema.exists((x: StructField) => x.name.equals(parameter) && x.dataType == StringType)
    })
    val newNameOfTextFeatures = textFeatures.map(setIndexesName)
    val numberFeature: Array[String] = features.filter(f => !textFeatures.contains(f))
    val newFeatures = numberFeature ++ newNameOfTextFeatures

    val textFeaturesIndexer: StringIndexerModel = new StringIndexer()
      .setInputCols(textFeatures)
      .setOutputCols(newNameOfTextFeatures)
      .fit(data)
    if (Context.saveModels)
      textFeaturesIndexer.write.overwrite().save(getTextFeatureIndexerFilePath)
    val newData: DataFrame = textFeaturesIndexer.transform(data)
    val vector = new VectorAssembler()
      .setInputCols(newFeatures)
      .setOutputCol("nonScaledFeatures")
      .transform(newData)
      .withColumnRenamed(label, "label")
      .select("nonScaledFeatures", "label")
    if (standardization == "MinMaxScaler") {
      new MinMaxScaler()
        .setInputCol("nonScaledFeatures")
        .setOutputCol("features")
        .fit(vector)
        .transform(vector)
        .select("features", "label")
    }
    else if (standardization == "StandardScaler") {
        new StandardScaler()
          .setWithStd(true)
          .setWithMean(true)
          .setInputCol("nonScaledFeatures")
          .setOutputCol("features")
          .fit(vector)
          .transform(vector)
          .select("features", "label")
    }
    else
      vector.withColumnRenamed("nonScaledFeatures", "features")
  }

}
